substrate/machine.py

```
@@ -0,0 +1,91 @@
+    IN = 1
+    OUT = 3
+    OPEN_DRAIN = 7
+    PULL_UP = 1
+    PULL_DOWN = 2
+    IRQ_RISING = 1
+    IRQ_FALLING = 2
+
+    _levels = {}      # id -> level (shared GPIO state, like the real chip)
+    _instances = {}   # id -> Pin (a given GPIO maps to one logical pin)
+
+    def __new__(cls, id, *args, **kwargs):
```

Ruff (0.15.18)

[error] Function argument `id` is shadowing a Python builtin

(A002)